ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -872,27 +896,48 @@ public CompletableFuture<Short> updateBlobExpirationAsync(BlobId blobId, long ex
     throw new UnsupportedOperationException("updateBlobExpirationAsync will not be implemented for AzureCloudDestinationSync");
   }
 
+  /**
+   * Get azure-blob-properties and wraps it in ambry cloud-blob-metadata object.
+   * @param blobId
+   * @return
+   * @throws CloudStorageException
+   */
+  @Override
+  public CloudBlobMetadata getCloudBlobMetadata(BlobId blobId) throws CloudStorageException {
+    AzureBlobLayoutStrategy.BlobLayout blobLayout = this.azureBlobLayoutStrategy.getDataBlobLayout(blobId);
+    try {
+      BlobProperties blobPropertiesCached = getBlobPropertiesCached(blobLayout);
+      if (blobPropertiesCached != null) {
+        return CloudBlobMetadata.fromMap(blobPropertiesCached.getMetadata());
+      }
+    } catch (CloudStorageException cse) {
+      throw cse;
+    } catch (Throwable t) {
+      throw new CloudStorageException(t.getMessage());
+    }
+    return null;
+  }
+
   @Override
   public Map<String, CloudBlobMetadata> getBlobMetadata(List<BlobId> blobIds) throws CloudStorageException {
-    /*
-        This is used by findMissingKeys, which doesn't even use the metadata provided.
-        It just discards it. But due to legacy reasons, we are required to impl this way.
-        We could use one-liner Azure SDK provided lightweight exists() method, which internally fetches blob-properties
-        and does the same thing that findMissingKeys does.
+    /**
+     * Just keeping this fn around for tests and legacy reasons but imo this is a horrible function.
+     * The fn takes a list as an arg, why not just a single blob ? It is only ever used to fetch metadata of a blob,
+     * not a list. And because of this design, the higher levels have to wrap the blob-id in a singleton-list.
+     * More importantly, the return type is map. The higher levels have to explicitly look for the result in the map
+     * with blob-id. If there is typo in the code in looking up the map, for eg. instead of storeKey.getId(),
+     * if we type storeKey() alone, then we make a mistake of thinking the blob is not found. Just stop using maps!
+     * Such lookup errors are easy to miss in code-reviews. And most legacy tests are broken beyond repair.
```

Comment:
Yeah, this is odd. If parameter is a list of BlobId, then the key of the map should be BlobId as well. Anyway, we are not caling getBlobMetadata with multiple blob ids anyway.

Reply:
Yeah, this fn created some bad code in the upper layers.